File: python3/binary_tree_preorder_traversal.py
from model import TreeNode

"""
Space   : O(1)
Time    : O(n)
"""

class Solution:
    def getPreorder(self, root: TreeNode) -> List[int]:
        res = []

        if root:
            res.append(root.val)
            res += self.getPreorder(root.left)
            res += self.getPreorder(root.right)

        return res

    def preorderTraversal(self, root: TreeNode) -> List[int]:
        return self.getPreorder(root)

File: golang/binary_tree_preorder_traversal.go
package main

import "golang/model"

/*
Space	: O(n)
Time	: O(n)
*/

func preorderTraversal(root *model.TreeNode) []int {
	res := []int{}
	if root != nil {
		res = append(res, root.Val)
		res = append(res, preorderTraversal(root.Left)...)
		res = append(res, preorderTraversal(root.Right)...)
	}

	return res
}
